### PR TITLE
Algorithm for clipping rectangles to rectangular windows

### PR DESCRIPTION
**Pull request by mcara**
This PR adds an algorithm for clipping multiple rectangles to clipping windows. Rectangles to be clipped must have sides parallel to coordinate axes. `get_clipped_pixels()` returns indices of the clipped pixels and their area.

This PR is to support commit https://github.com/spacetelescope/jwst/commit/af1d903e1aeff917397e86b65e26b8733df8f2c8 in #5508

CC: @hbushouse @nden

### CODE REVIEW

**Comment by hbushouse** [comment]
The new algorithm itself seems to be fine to me, but there are obviously errors in the package setup, given that the test module is not able to import `winclip`. These seem to be similar to the problems I had when trying to get setup to build things properly when including a C module. I can get it to build OK in my local checkout, so that I can run the code, but could never find the right way to get it to build properly in the CI systems.

File: setup.cfg
Comment (jdavies-st) [inline setup.cfg:29]: This is not really needed here, because by the time `setup` is run, numpy needs to be installed already.  Instead you need to list it in pyproject.toml, and use `oldest-supported-numpy` instead, as that's what will be used to build the C code.

See

https://github.com/astropy/astropy/blob/5f8866e41164293e463a3cc53fadb5566ba488cc/pyproject.toml#L7

for an example.

File: setup.py
Comment (jdavies-st) [inline setup.py:93]: These are not necessary (nor any of the code above defining the global variables), as pyproject.toml makes sure that `numpy` is installed before `setup()` is run.  And the meta already gets pulled from setup.cfg, as specified in pyproject.toml.

File: setup.py
Comment (jdavies-st) [inline setup.py:89]: I believe none of this except `include_dirs` is needed, as again, pyproject.toml assures that `numpy` is installed before `setup()` is run.  Also we don't support Windows (currently).

File: setup.py
Comment (mcara) [reply setup.py:93]: But if this is not incorrect, i.e., it does not hurt, I would prefer to keep this code in.

As a test, I created a new/clean Python environment and I commented out the code that you say is optional and tried to install `jwst` into this environment. Here is the error that I get:

```python
(clnenv) dacia:jwst mcara$ pip install .
Processing /.../jwst
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  ERROR: Command errored out with exit status 1:
   command: /.../clnenv/bin/python3 /.../clnenv/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py get_requires_for_build_wheel /var/folders/kb/78_n024d2jb8p24bf6t2c3_c0001_0/T/tmpbnsu8zik
       cwd: /private/var/folders/kb/78_n024d2jb8p24bf6t2c3_c0001_0/T/pip-req-build-ul60c3p9
  Complete output (16 lines):
  Traceback (most recent call last):
    File "/.../clnenv/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 280, in <module>
      main()
    File "/.../clnenv/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 263, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/.../clnenv/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py", line 114, in get_requires_for_build_wheel
      return hook(config_settings)
    File "/private/var/folders/kb/78_n024d2jb8p24bf6t2c3_c0001_0/T/pip-build-env-5c_84cv7/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 154, in get_requires_for_build_wheel
      return self._get_build_requires(
    File "/private/var/folders/kb/78_n024d2jb8p24bf6t2c3_c0001_0/T/pip-build-env-5c_84cv7/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 135, in _get_build_requires
      self.run_setup()
    File "/private/var/folders/kb/78_n024d2jb8p24bf6t2c3_c0001_0/T/pip-build-env-5c_84cv7/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 150, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 77, in <module>
      import numpy
  ModuleNotFoundError: No module named 'numpy'
  ----------------------------------------
```

Should users be able to install `jwst` by clonning it and then running `pip install .`?

File: setup.py
Comment (jdavies-st) [inline setup.py:26]: This is all handled from pyproject.toml build backend, and already currently works. I.e. you don't need to parse setup.cfg.  This already happens because

```
build-backend = "setuptools.build_meta"
```

File: setup.py
Comment (jdavies-st) [reply setup.py:93]: Yes, if you have numpy listed in `pyproject.toml`, then it _will_ work.  This is the modern way to do it.

Comment (jdavies-st) [reply setup.py:93]: I.e. the only reason you are getting to the point of not having `numpy` in the above traceback is because it is not listed in pyproject.toml as it needs to be.  `jwst` is PEP517-compliant, and this is how one guarantees that the build requirements are installed.

Comment (mcara) [reply setup.py:93]: @jdavies-st Alright. I moved `'numpy'` requirement to `pyproject.toml` and now `pip install .` works but the now `python setup.py install` fails:

```python
jwst mcara$ python setup.py install
WARNING: The wheel package is not available.
Traceback (most recent call last):
  File "setup.py", line 78, in <module>
    import numpy
ModuleNotFoundError: No module named 'numpy'
```

Comment (jdavies-st) [reply setup.py:93]: You need `import numpy` at the top of `setup.py`.  pyproject.toml will guarantee it is installed in the isolated build environment, but you'll still need to import it.

Be sure to use `oldest-supported-numpy` in pyproject.toml, so that we avoid the problems we saw a few months ago when the numpy folks delevered 1.20 and we were still using 1.19 in our deliveries to SDP and it broke the installs.

**Comment by mcara** [comment]
@hbushouse When testing this PR with #5508 (or the other way) you may want to uncomment the following lines in `winclip.c`: https://github.com/spacetelescope/jwst/blob/24b647240aad5d5be9a7e6466a43819568e5fb41/jwst/lib/src/winclip.c#L186-L193
in order to simulate the accuracy loss in the original code [to get results closer to the original code].

File: setup.py
Comment (mcara) [reply setup.py]: @jdavies-st OK. Removed support for installing/compiling on Windows and support for installing using `python setup.py install` in https://github.com/spacetelescope/jwst/pull/5978/commits/a9de43db2bba745cd793f57d20023c7d1b58477a

**Comment by mcara** [comment]
I have no idea how to get this PR to pass this rigorous `codecov` test.

File: tox.ini
Comment (pllim) [inline tox.ini:44]: If you find a solution, please circle back to spacetelescope/synphot_refactor#295 . Thanks!

**Comment by pllim** [comment]
I don't think codecov measures C-extension accurately. 🤷

File: tox.ini
Comment (mcara) [reply tox.ini:44]: I think I found one for `jwst` but it does not work for `synphot_refactor` yet (one failing test). I'll make another PR for `jwst` pipeline first and and then if you like it - you can try it on `synphot_refactor`.

The biggest issue that's killing this idea for `synphot_refactor` is your import of `_astropy_init` (due to affiliated package status).

**Comment by mcara** [comment]
> I don't think codecov measures C-extension accurately. 🤷

No, hire coverage measures coverage of `setup.py` and somehow some imports are not covered.

File: tox.ini
Comment (pllim) [reply tox.ini:44]: Why would `_astropy_init` break `tox`? 🤔

**Comment by jdavies-st** [comment]
> here coverage measures coverage of `setup.py` and somehow some imports are not covered.

It should not be covering `setup.py`.  That's my mistake in our coverage setup.  Feel free to add it setup.py to the omits:

https://github.com/spacetelescope/jwst/blob/0d5269d2760fcd2966fe8480cee10da4492e8939/setup.cfg#L137-L151

File: tox.ini
Comment (jdavies-st) [inline tox.ini:72]: Commenting out the `changedir` directive will break the purpose of doing `pytest --pyargs`, which is to test the package on the installed version of the package in `site-packages`.  `pip install .` should be installing the build C lib in `site-packages` here.

**Review by hbushouse — APPROVED** [review]
Approving now, with the understanding that issues concerning the "correct" way to build C extensions in setup.py and tox.ini will be resolved independently in the near future after this is merged.

**Pull request merged**
